File: modules/lab3d/scene/Camera.lua
--[[---------------------------------------------------------------------------
	Camera module.
	Provides lab3d.scene.ICamera interface service.
--]]---------------------------------------------------------------------------

local glm = require "glm"
local View = require "lab3d.dom.View"

-------------------------------------------------------------------------------
-- Component declaration
-------------------------------------------------------------------------------
local Camera =  co.Component( "lab3d.scene.Camera" )


function Camera:__init()
	self.znear = self.znear or 100
	self.zfar = self.zfar or 10000
	self.fovy = self.fovy or 60
	self.aspect = self.aspect or 1
	self.viewportWidth = self.viewportW or 800
	self.viewportHeight = self.viewportH or 600
	local viewObj = View{ position = glm.Vec3( 0, -100, 0 ) }
	self.view = viewObj.view; -- default camera view
end

function Camera:setView( view )
	self.view = view
end

function Camera:getView()
	return self.view
end

function Camera:getFovy()
	return self.fovy
end

function Camera:setFovy( fovy )
	self.fovy = fovy
end

function Camera:getAspect()
	return self.aspect
end

function Camera:setAspect( aspect )
	self.aspect = aspect
end

function Camera:getZNear()
	return self.znear
end

function Camera:setZNear( znear )
	self.znear = znear
end

function Camera:getZFar()
	return self.zfar
end

function Camera:setZFar( zfar )
	self.zfar = zfar
end

function Camera:getViewportWidth()
	return self.viewportWidth
end

function Camera:setViewportWidth( width )
	self.viewportWidth = width
end

function Camera:setViewportHeight( height )
	self.viewportHeight = height
end

function Camera:getViewportHeight( height )
	return self.viewportHeight
end

return Camera

File: src/lab3d.dom/Camera.cpp
#include "Camera_Base.h"
#include <co/Coral.h>
#include <co/RefPtr.h>
#include <lab3d/dom/IView.h>

namespace lab3d {
namespace dom {

class Camera : public Camera_Base
{
public:
	Camera() : _fovy( 60 ), _zNear( 1 ), _zFar( 10000 ),
		_viewportWidth( 800 ), _viewportHeight( 600 )
	{
		_view = co::newInstance( "lab3d.dom.View" )->getService<lab3d::dom::IView>();
		updateAspect();
	}

	virtual ~Camera()
	{
		// empty destructor
	}

	// ------ lab3d.dom.ICamera Methods ------ //
	
	lab3d::dom::IView* getView()
	{
		return _view.get();
	}
	
	void setView( lab3d::dom::IView* view )
	{
		_view = view;
	}
	
	double getFovy()
	{
		return _fovy;
	}
	
	void setFovy( double fovy )
	{
		_fovy = fovy;
	}
	
	double getZNear()
	{
		return _zNear;
	}
	
	void setZNear( double zNear )
	{
		_zNear = zNear;
	}
	
	double getZFar()
	{
		return _zFar;
	}
	
	void setZFar( double zFar )
	{
		_zFar = zFar;
	}
	
	co::int32 getViewportWidth()
	{
		return _viewportWidth;
	}
	
	void setViewportWidth( co::int32 viewportWidth )
	{
		_viewportWidth = viewportWidth;
		updateAspect();
	}

	co::int32 getViewportHeight()
	{
		return _viewportHeight;
	}

	void setViewportHeight( co::int32 viewportHeight )
	{
		_viewportHeight = viewportHeight;
		updateAspect();
	}
	
	double getAspect()
	{
		return _aspect;
	}
	
	void setAspect( double aspect )
	{
		_aspect = aspect;
	}

private:
	inline void updateAspect()
	{
		_aspect = static_cast<double>( _viewportWidth ) / _viewportHeight;
	}

private:
	co::RefPtr<lab3d::dom::IView> _view;
	double _fovy;
	double _zNear;
	double _zFar;
	double _aspect;
	co::int32 _viewportWidth;
	co::int32 _viewportHeight;
};

CORAL_EXPORT_COMPONENT( Camera, Camera );

} // namespace dom
} // namespace lab3d
